### PR TITLE
Replace recordingSpan attributes implementation with map of attributes

### LINKED ISSUE #2554: Span attribute drop logic does not comply with the specification.
### Description

Currently the trace SDK drops attributes based on a least-recently-used algorithm: the most recently unique attributes added are kept and the oldest not updated attributes are dropped when capacity is reached. However, [the specification states](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/common/common.md):

> for each unique attributes key, addition of which would result in exceeding the limit, SDK MUST discard that key/value pair

Meaning if a user adds an attribute after capacity is reached that attribute (if it has a unique key) should be dropped. This is not the behavior currently implemented.

### Environment

- opentelemetry-go version: v1.3.0

### Steps To Reproduce

```go
package trace // import "go.opentelemetry.io/otel/sdk/trace"

import (
	"context"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"go.opentelemetry.io/otel/attribute"
)

func TestSpanAttributeCapacityDropOrder(t *testing.T) {
	// The tracing specification states:
	//
	//   For each unique attribute key, addition of which would result in
	//   exceeding the limit, SDK MUST discard that key/value pair
	//
	// Therefore, adding attributes after the capacity is reached should
	// result in those attributes being dropped.

	const (
		instName = "TestSpanAttributeCapacityDropOrder"
		spanName = "test span"
	)

	te := NewTestExporter()
	tp := NewTracerProvider(
		WithSyncer(te),
		WithSpanLimits(SpanLimits{AttributeCountLimit: 1}),
	)
	attrs := []attribute.KeyValue{
		attribute.String("key1", "value1"),
		attribute.String("key2", "value2"),
	}

	_, span := tp.Tracer(instName).Start(context.Background(), spanName)
	span.SetAttributes(attrs[0])
	// These should be dropped based on limits.
	span.SetAttributes(attrs[1])
	span.SetAttributes(attrs...)
	span.End()

	got, ok := te.GetSpan(spanName)
	require.Truef(t, ok, "span %s not exported", spanName)
	assert.Equal(t, attrs[:1], got.attributes)
}
```

### Expected behavior

The included test should pass and the first attributes a user passes should be preserved.

### PR DESCRIPTION
**Pull request by MrAlias**
Instead of an LRU strategy for cap-ing span attributes, comply with the specification and drop last added. This means the `attributesMap` type can be replaced with a simple map.

Resolves #2554 and relates to #2402

### Testing Performance

This looks to have considerable memory and allocation reduction and faster computations.

```
$ benchstat bench-main-53ead308 bench-branch

name                                          old time/op    new time/op    delta
SpanProcessor-8                                 14.9µs ±10%    12.6µs ± 9%  -15.45%  (p=0.000 n=8+10)
SpanProcessorVerboseLogging-8                   15.5µs ± 3%    12.8µs ±16%  -17.45%  (p=0.000 n=8+10)
StartEndSpan/AlwaysSample-8                      978ns ±14%     708ns ±24%  -27.66%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-8                       367ns ± 2%     312ns ± 9%  -15.14%  (p=0.000 n=8+9)
SpanSetAttributesOverCapacity-8                 23.1µs ± 5%     2.3µs ±13%  -89.97%  (p=0.000 n=10+9)
SpanWithAttributes_4/AlwaysSample-8             1.85µs ± 5%    1.28µs ±11%  -31.00%  (p=0.000 n=8+9)
SpanWithAttributes_4/NeverSample-8               513ns ± 7%     481ns ± 5%   -6.25%  (p=0.004 n=9+9)
SpanWithAttributes_8/AlwaysSample-8             2.68µs ± 5%    1.60µs ± 5%  -40.35%  (p=0.000 n=8+9)
SpanWithAttributes_8/NeverSample-8               679ns ± 5%     644ns ± 7%   -5.15%  (p=0.021 n=8+10)
SpanWithAttributes_all/AlwaysSample-8           2.09µs ± 7%    1.36µs ± 3%  -35.29%  (p=0.000 n=9+9)
SpanWithAttributes_all/NeverSample-8             566ns ± 5%     508ns ± 7%  -10.36%  (p=0.000 n=8+9)
SpanWithAttributes_all_2x/AlwaysSample-8        3.72µs ±12%    2.57µs ± 9%  -30.98%  (p=0.000 n=10+9)
SpanWithAttributes_all_2x/NeverSample-8          753ns ± 6%     690ns ± 8%   -8.47%  (p=0.001 n=8+9)
SpanWithEvents_4/AlwaysSample-8                 2.04µs ± 7%    1.55µs ± 9%  -23.98%  (p=0.000 n=10+9)
SpanWithEvents_4/NeverSample-8                   386ns ± 3%     324ns ± 7%  -16.00%  (p=0.000 n=9+9)
SpanWithEvents_8/AlwaysSample-8                 3.02µs ± 3%    2.32µs ± 8%  -23.28%  (p=0.000 n=9+8)
SpanWithEvents_8/NeverSample-8                   395ns ± 8%     339ns ± 7%  -14.04%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/AlwaysSample-8    1.26µs ± 8%    1.01µs ±12%  -20.06%  (p=0.000 n=10+9)
SpanWithEvents_WithStackTrace/NeverSample-8      423ns ±10%     378ns ±16%  -10.57%  (p=0.008 n=9+10)
SpanWithEvents_WithTimestamp/AlwaysSample-8     1.24µs ± 5%    0.98µs ± 9%  -21.43%  (p=0.000 n=10+9)
SpanWithEvents_WithTimestamp/NeverSample-8       470ns ± 2%     418ns ± 8%  -11.07%  (p=0.000 n=10+10)
TraceID_DotString-8                             82.2ns ± 5%    84.2ns ±10%     ~     (p=0.278 n=10+9)
SpanID_DotString-8                              62.3ns ± 2%    62.7ns ± 6%     ~     (p=0.859 n=10+9)

name                                          old alloc/op   new alloc/op   delta
SpanProcessor-8                                 12.3kB ± 0%     8.6kB ± 0%  -29.99%  (p=0.000 n=10+10)
SpanProcessorVerboseLogging-8                   13.4kB ± 0%     9.7kB ± 0%  -27.43%  (p=0.000 n=10+9)
StartEndSpan/AlwaysSample-8                       816B ± 0%      448B ± 0%  -45.10%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-8                        224B ± 0%      128B ± 0%  -42.86%  (p=0.000 n=10+10)
SpanSetAttributesOverCapacity-8                 15.4kB ± 0%     1.0kB ± 0%  -93.33%  (p=0.000 n=10+10)
SpanWithAttributes_4/AlwaysSample-8             1.73kB ± 0%    1.28kB ± 0%  -25.93%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-8                480B ± 0%      384B ± 0%  -20.00%  (p=0.000 n=10+10)
SpanWithAttributes_8/AlwaysSample-8             2.43kB ± 0%    1.54kB ± 0%  -36.84%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-8                736B ± 0%      640B ± 0%  -13.04%  (p=0.000 n=10+10)
SpanWithAttributes_all/AlwaysSample-8           1.90kB ± 0%    1.34kB ± 0%  -29.41%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-8              544B ± 0%      448B ± 0%  -17.65%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/AlwaysSample-8        3.20kB ± 0%    2.83kB ± 0%  -11.74%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/NeverSample-8           864B ± 0%      768B ± 0%  -11.11%  (p=0.000 n=10+10)
SpanWithEvents_4/AlwaysSample-8                 1.50kB ± 0%    0.88kB ± 0%  -41.49%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-8                    224B ± 0%      128B ± 0%  -42.86%  (p=0.000 n=10+10)
SpanWithEvents_8/AlwaysSample-8                 2.21kB ± 0%    1.33kB ± 0%  -39.86%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-8                    224B ± 0%      128B ± 0%  -42.86%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/AlwaysSample-8      992B ± 0%      560B ± 0%  -43.55%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-8       240B ± 0%      144B ± 0%  -40.00%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/AlwaysSample-8     1.02kB ± 0%    0.58kB ± 0%  -42.52%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-8        264B ± 0%      168B ± 0%  -36.36%  (p=0.000 n=10+10)

name                                          old allocs/op  new allocs/op  delta
SpanProcessor-8                                   85.0 ± 0%      45.0 ± 0%  -47.06%  (p=0.000 n=10+10)
SpanProcessorVerboseLogging-8                     91.0 ± 0%      51.0 ± 0%  -43.96%  (p=0.000 n=10+9)
StartEndSpan/AlwaysSample-8                       7.00 ± 0%      3.00 ± 0%  -57.14%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-8                        3.00 ± 0%      2.00 ± 0%  -33.33%  (p=0.000 n=10+10)
SpanSetAttributesOverCapacity-8                    264 ± 0%         4 ± 0%  -98.48%  (p=0.000 n=10+10)
SpanWithAttributes_4/AlwaysSample-8               17.0 ± 0%       5.0 ± 0%  -70.59%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-8                4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.000 n=10+10)
SpanWithAttributes_8/AlwaysSample-8               25.0 ± 0%       5.0 ± 0%  -80.00%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-8                4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.000 n=10+10)
SpanWithAttributes_all/AlwaysSample-8             19.0 ± 0%       5.0 ± 0%  -73.68%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-8              4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/AlwaysSample-8          30.0 ± 0%       6.0 ± 0%  -80.00%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/NeverSample-8           4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.000 n=10+10)
SpanWithEvents_4/AlwaysSample-8                   18.0 ± 0%      10.0 ± 0%  -44.44%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-8                    3.00 ± 0%      2.00 ± 0%  -33.33%  (p=0.000 n=10+10)
SpanWithEvents_8/AlwaysSample-8                   27.0 ± 0%      15.0 ± 0%  -44.44%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-8                    3.00 ± 0%      2.00 ± 0%  -33.33%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/AlwaysSample-8      11.0 ± 0%       6.0 ± 0%  -45.45%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-8       4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/AlwaysSample-8       12.0 ± 0%       7.0 ± 0%  -41.67%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-8        5.00 ± 0%      4.00 ± 0%  -20.00%  (p=0.000 n=10+10)

```

### CODE REVIEW

**Comment by MrAlias** [comment]
`benchstat` comparison against `main`:

```
name                                          old time/op    new time/op    delta
SpanProcessor-8                                 15.7µs ± 5%    13.7µs ±20%  -13.15%  (p=0.003 n=9+9)
SpanProcessorVerboseLogging-8                   16.1µs ± 3%    14.4µs ± 3%  -10.35%  (p=0.000 n=9+7)
StartEndSpan/AlwaysSample-8                     1.05µs ± 2%    0.84µs ± 9%  -19.81%  (p=0.000 n=9+10)
StartEndSpan/NeverSample-8                       389ns ± 5%     367ns ± 6%   -5.55%  (p=0.000 n=10+10)
SpanWithAttributes_4/AlwaysSample-8             2.00µs ± 9%    1.42µs ± 7%  -29.25%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-8               548ns ± 9%     527ns ± 9%   -3.85%  (p=0.024 n=9+9)
SpanWithAttributes_8/AlwaysSample-8             2.86µs ± 5%    1.78µs ± 7%  -37.79%  (p=0.000 n=9+9)
SpanWithAttributes_8/NeverSample-8               702ns ± 6%     692ns ± 8%     ~     (p=0.604 n=9+10)
SpanWithAttributes_all/AlwaysSample-8           2.32µs ±15%    1.52µs ± 3%  -34.45%  (p=0.000 n=10+8)
SpanWithAttributes_all/NeverSample-8             584ns ± 8%     571ns ± 3%     ~     (p=0.203 n=10+8)
SpanWithAttributes_all_2x/AlwaysSample-8        3.80µs ± 6%    2.72µs ±13%  -28.46%  (p=0.000 n=9+9)
SpanWithAttributes_all_2x/NeverSample-8          758ns ± 5%     747ns ± 4%     ~     (p=0.436 n=9+9)
SpanWithEvents_4/AlwaysSample-8                 2.19µs ± 3%    2.00µs ±13%   -8.76%  (p=0.000 n=9+10)
SpanWithEvents_4/NeverSample-8                   395ns ± 4%     381ns ± 1%   -3.57%  (p=0.029 n=9+7)
SpanWithEvents_8/AlwaysSample-8                 3.32µs ±11%    2.94µs ± 1%  -11.41%  (p=0.000 n=10+9)
SpanWithEvents_8/NeverSample-8                   404ns ± 4%     399ns ± 5%     ~     (p=0.105 n=8+8)
SpanWithEvents_WithStackTrace/AlwaysSample-8    1.36µs ± 6%    1.15µs ± 5%  -15.38%  (p=0.000 n=9+9)
SpanWithEvents_WithStackTrace/NeverSample-8      435ns ± 6%     417ns ± 2%   -4.07%  (p=0.000 n=9+9)
SpanWithEvents_WithTimestamp/AlwaysSample-8     1.38µs ± 5%    1.17µs ± 1%  -15.51%  (p=0.000 n=10+8)
SpanWithEvents_WithTimestamp/NeverSample-8       492ns ± 3%     474ns ± 5%   -3.59%  (p=0.001 n=9+9)
TraceID_DotString-8                             84.7ns ± 4%    82.3ns ± 2%   -2.88%  (p=0.010 n=10+7)
SpanID_DotString-8                              64.2ns ± 5%    62.5ns ± 3%     ~     (p=0.079 n=10+9)

name                                          old alloc/op   new alloc/op   delta
SpanProcessor-8                                 12.6kB ± 0%    11.8kB ± 0%   -6.35%  (p=0.000 n=9+10)
SpanProcessorVerboseLogging-8                   13.7kB ± 0%    12.9kB ± 0%   -5.82%  (p=0.000 n=10+9)
StartEndSpan/AlwaysSample-8                       848B ± 0%      768B ± 0%   -9.43%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-8                        224B ± 0%      224B ± 0%     ~     (all equal)
SpanWithAttributes_4/AlwaysSample-8             1.76kB ± 0%    1.60kB ± 0%   -9.09%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-8                480B ± 0%      480B ± 0%     ~     (all equal)
SpanWithAttributes_8/AlwaysSample-8             2.46kB ± 0%    1.86kB ± 0%  -24.68%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-8                736B ± 0%      736B ± 0%     ~     (all equal)
SpanWithAttributes_all/AlwaysSample-8           1.94kB ± 0%    1.66kB ± 0%  -14.05%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-8              544B ± 0%      544B ± 0%     ~     (all equal)
SpanWithAttributes_all_2x/AlwaysSample-8        3.24kB ± 0%    3.15kB ± 0%   -2.72%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/NeverSample-8           864B ± 0%      864B ± 0%     ~     (all equal)
SpanWithEvents_4/AlwaysSample-8                 1.54kB ± 0%    1.46kB ± 0%   -5.21%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-8                    224B ± 0%      224B ± 0%     ~     (all equal)
SpanWithEvents_8/AlwaysSample-8                 2.24kB ± 0%    2.16kB ± 0%   -3.57%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-8                    224B ± 0%      224B ± 0%     ~     (all equal)
SpanWithEvents_WithStackTrace/AlwaysSample-8    1.02kB ± 0%    0.94kB ± 0%   -7.81%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-8       240B ± 0%      240B ± 0%     ~     (all equal)
SpanWithEvents_WithTimestamp/AlwaysSample-8     1.05kB ± 0%    0.97kB ± 0%   -7.63%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-8        264B ± 0%      264B ± 0%     ~     (all equal)

name                                          old allocs/op  new allocs/op  delta
SpanProcessor-8                                    105 ± 0%        65 ± 0%  -38.10%  (p=0.000 n=10+10)
SpanProcessorVerboseLogging-8                      111 ± 0%        71 ± 0%  -36.04%  (p=0.000 n=10+10)
StartEndSpan/AlwaysSample-8                       9.00 ± 0%      5.00 ± 0%  -44.44%  (p=0.000 n=10+10)
StartEndSpan/NeverSample-8                        3.00 ± 0%      3.00 ± 0%     ~     (all equal)
SpanWithAttributes_4/AlwaysSample-8               19.0 ± 0%       7.0 ± 0%  -63.16%  (p=0.000 n=10+10)
SpanWithAttributes_4/NeverSample-8                4.00 ± 0%      4.00 ± 0%     ~     (all equal)
SpanWithAttributes_8/AlwaysSample-8               27.0 ± 0%       7.0 ± 0%  -74.07%  (p=0.000 n=10+10)
SpanWithAttributes_8/NeverSample-8                4.00 ± 0%      4.00 ± 0%     ~     (all equal)
SpanWithAttributes_all/AlwaysSample-8             21.0 ± 0%       7.0 ± 0%  -66.67%  (p=0.000 n=10+10)
SpanWithAttributes_all/NeverSample-8              4.00 ± 0%      4.00 ± 0%     ~     (all equal)
SpanWithAttributes_all_2x/AlwaysSample-8          32.0 ± 0%       8.0 ± 0%  -75.00%  (p=0.000 n=10+10)
SpanWithAttributes_all_2x/NeverSample-8           4.00 ± 0%      4.00 ± 0%     ~     (all equal)
SpanWithEvents_4/AlwaysSample-8                   20.0 ± 0%      16.0 ± 0%  -20.00%  (p=0.000 n=10+10)
SpanWithEvents_4/NeverSample-8                    3.00 ± 0%      3.00 ± 0%     ~     (all equal)
SpanWithEvents_8/AlwaysSample-8                   29.0 ± 0%      25.0 ± 0%  -13.79%  (p=0.000 n=10+10)
SpanWithEvents_8/NeverSample-8                    3.00 ± 0%      3.00 ± 0%     ~     (all equal)
SpanWithEvents_WithStackTrace/AlwaysSample-8      13.0 ± 0%       9.0 ± 0%  -30.77%  (p=0.000 n=10+10)
SpanWithEvents_WithStackTrace/NeverSample-8       4.00 ± 0%      4.00 ± 0%     ~     (all equal)
SpanWithEvents_WithTimestamp/AlwaysSample-8       14.0 ± 0%      10.0 ± 0%  -28.57%  (p=0.000 n=10+10)
SpanWithEvents_WithTimestamp/NeverSample-8        5.00 ± 0%      5.00 ± 0%     ~     (all equal)
```

**Review by Aneurysm9 — APPROVED** [review]

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:242]: That's a good question. Looking at the specification it is unspecified how to count dropped attributes. The only thing I can find is how they are reported, not what constitutes a dropped attribute.

I was also wondering if we should be counting attributes dropped because they are invalid.

Given the unspecified state of dropped attributes I think we are free to make the decision ourselves how we want to count them. I think we should probably change the current strategy to be all attributes dropped (every invalid or duplicate) or only valid unique ones. I think the former strategy is going to better server users. Dropping attributes and not reporting them will result in lost information and may confuse users to think there are bugs in opentelemetry-go or their code. Reporting these attributes as dropped will still obscure the reason they are dropped, but it will at least report something about the action.

Comment (MadVikingGod) [reply sdk/trace/span.go:242]: I had the same question, but I also realized we would have to capture the attribute keys somewhere to be able to check this uniqueness which would in most cases negate the reason we are dropping.

File: sdk/trace/trace_test.go
Comment (MadVikingGod) [inline sdk/trace/trace_test.go:396]: Why did you change this test?

File: sdk/trace/trace_test.go
Comment (MadVikingGod) [inline sdk/trace/trace_test.go:497]: Can we also assert the DroppedAttributes() == 1?
```suggestion
	assert.Contains(t, got.attributes, attrs[1])
	assert.Contains(t, got.attributes, attrs[2])
	asster.Equal(t, got.droppedAttributes, 1)
```

**Review by MadVikingGod — APPROVED** [review]

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:243]: I've updated to count all attributes dropped (invalid or duplicate).

File: sdk/trace/trace_test.go
Comment (MrAlias) [reply sdk/trace/trace_test.go:396]: Because there are two attributes set originally, the order of those attributes is a something checked indirectly. That order is not guaranteed anymore.

This updates the test to still set two attributes, similar to the original, but it updates the key-value. Resulting in a single attribute for the exported span. This update behavior is something guaranteed and is functionality also required by the specification.

File: sdk/trace/span.go
Comment (jmacd) [inline sdk/trace/span.go:147]: Could we refer to `tr.provider.spanLimits.AttributeCountLimit` directly?

File: sdk/trace/tracer.go
Comment (jmacd) [inline sdk/trace/tracer.go:132]: I see here that the span stores `tr`. Similar to my first question, can we eliminate `maxAttributes`, `resource`, and `spanLimits` fields? These are call copied from the tracer, but we keep a reference to the tracer so we may not need these copies.

**Review by jmacd — APPROVED** [review]
This is a significant improvement, approving on the merits. For myself I'd argue to go even further.

I would optimize for StartSpan performance => keep a reference to the original `[]attribute.KeyValue` and document that the caller MUST not modify that slice. If the slice length exceeds the maximum, call `deduplicateAndDrop()`. Do not check for duplicates, otherwise, just store the array.

The `SetAttributes()` method would be optimized to avoid deduplication in a hot path as well, if possible. As long as the total slice length will not exceed the limit, simply `append(sp.attributes, newAttributes...)` and defer the deduplication. If the total slice length exceeds the limit, call `deduplicateAndDrop()`.

The call to `attributesLocked()` is when I would deduplicate. We don't need to allocate a map to deduplicate, we can sort in place, so the end result will be that if you only call `StartSpan` and never call `SetAttributes`, the SDK itself won't perform any allocations for span attributes.

**Comment by Aneurysm9** [comment]
> I would optimize for StartSpan performance => keep a reference to the original `[]attribute.KeyValue` and document that the caller MUST not modify that slice.

I think that would be a significant change to the data ownership semantics and could break existing uses.  I would not support making such a change without a major version bump.

Making that change in the SDK also could lead to a situation where different SDK implementations behave differently and code that is correct against one implementation could be very incorrect against another.  That seems like something we should try to avoid.

File: sdk/trace/span.go
Comment (MrAlias) [reply sdk/trace/span.go:147]: SGTM, I'll push a change.

File: sdk/trace/tracer.go
Comment (MrAlias) [reply sdk/trace/tracer.go:132]: I'm happy to include changes to how `maxAttributes` is represented in this PR, but will hold off changes to `resource` and `spanLimits` for another PR dedicated to these changes.

Comment (MrAlias) [reply sdk/trace/tracer.go]: https://github.com/open-telemetry/opentelemetry-go/pull/2575

File: sdk/trace/trace_test.go
Comment (MadVikingGod) [reply sdk/trace/trace_test.go:396]: That makes me think that our assert isn't comprehensive enough for our test cases.

But, we cover this behavior in another test, so I won't hold this up.

**Review by MadVikingGod — APPROVED** [review]

**Comment by MrAlias** [comment]
> > I would optimize for StartSpan performance => keep a reference to the original `[]attribute.KeyValue` and document that the caller MUST not modify that slice.
> 
> I think that would be a significant change to the data ownership semantics and could break existing uses. I would not support making such a change without a major version bump.
> 
> Making that change in the SDK also could lead to a situation where different SDK implementations behave differently and code that is correct against one implementation could be very incorrect against another. That seems like something we should try to avoid.

I agree a change to the data ownership semantics is probably not something we want to do.

However, I am intrigued at the possibility to use a slice here. If we can realize the same performance improvements shown here and potentially maintain order of the attributes according to how the user adds them it seems like something worth exploring. I put together an initial implementation of this and it looks worth developing into a full alternate approach and evaluating the merits. I will do so before merging this.

**Comment by MadVikingGod** [comment]
I just noticed that with the removal of the attributeMap and its tests we have lost coverage that `(&recordingSpan).Attributes() == nil`.  Same that `DroppedCount() == 0`.  As long as this is the behavior already I'm still ok with merging this, just needs an issue to go back and add tests.

**Comment by MrAlias** [comment]
> I just noticed that with the removal of the attributeMap and its tests we have lost coverage that `(&recordingSpan).Attributes() == nil`. Same that `DroppedCount() == 0`. As long as this is the behavior already I'm still ok with merging this, just needs an issue to go back and add tests.

I don't see the existing test coverage, but I'm happy to add these tests.

**Comment by MrAlias** [comment]
Reverting to a draft so this doesn't accidentally get merged. I plan to present this and https://github.com/open-telemetry/opentelemetry-go/pull/2576 at the SIG meeting today and discuss which approach we should move forward with.

**Comment by MrAlias** [comment]
Closing in favor of https://github.com/open-telemetry/opentelemetry-go/pull/2576

**Pull request closed without merging**
